libtiledbsoma/src/pyapi/libtiledbsoma.cc

```diff
@@ -322,18 +408,113 @@ PYBIND11_MODULE(libtiledbsoma, m) {
                     uintptr_t arrow_array_ptr = (uintptr_t)(&arrow_array);
 
                     // Call array._export_to_c to get arrow array and schema
+                    //
+                    // If ever a NumPy array gets in here, there will be an
+                    // exception like "AttributeError: 'numpy.ndarray' object
+                    // has no attribute '_export_to_c'".
+                    // TODO: try/catch with something more helpful of the form
+                    // "Maybe not an Arrow array".
                     array.attr("_export_to_c")(
                         arrow_array_ptr, arrow_schema_ptr);
 
-                    // 'l' = arrow data type int64
                     if (!strcmp(arrow_schema.format, "l")) {
                         tcb::span<int64_t> data{
                             (int64_t*)arrow_array.buffers[1],
```

Comment:
We have been using `arrow_array.buffers[1]` as the data buffer. It would be safer to use something like the code below, because the data buffer will be in `arrow_array.buffers[2]` if offsets are present.

```c++
int data_index = arrow_array.n_buffers - 1;
...
(int64_t*)arrow_array.buffers[data_index],
```

Existing if/else logic may already avoid using the wrong buffer index, but explicitly avoiding the issue is easy to add.

Reply:
@gspowley oops, I merged with this issue unresolved. Putting up a follow-up PR now ...

Reply:
@gspowley #986